### PR TITLE
Improve index-neutrality of `LinearAlgebra`

### LINKED ISSUE #17513: _lu() assumes that the domain for the array it is provided will also be zero based
### Summary of Problem

The _lu() helper function by default assumes all the arrays provided to it are 0 based.

### Steps to Reproduce

**Source Code:**

```chapel
use LinearAlgebra;
var dom = {1..2,1..2};
var arr : [dom] real = 1;
var x = lu(arr);
writeln(x(0));
```
since here indices of ```arr``` is 1 based and not 0, it gives a run-time error ```index out of bounds```

### PR DESCRIPTION
**Pull request by Guillaume-Helbecque**
This PR is part of #17624 and improves the index-neutrality (any offset) of the following procedures:
- `solve_tril`
- `solve_triu`
- `solve`
    - private `_lu`
    - private `permute`
- `lu` (implicitly)
- `det` (implicitly)

It also adds several correctness tests.

Resolves #17513

### CODE REVIEW

**Comment by Guillaume-Helbecque** [comment]
The correctness tests added in this PR are derived from the long-time pending PR #17558.

**Comment by Guillaume-Helbecque** [comment]
**Note:** Working on this PR, I identified some issues that could be solved in follow up PRs:

- `solve_*` methods doesn't work when arrays contain `int`, due to integral division;
- I wonder why the `solve_tril` method has a `unit_diag` argument, while the `solve_triu` doesn't have. It seems that the latter has been introduced for performance reason, but removing it should make no difference. In addition, `solve_tril` is mainly called inside `solve`, which doesn't set `unit_diag`;
- We could benefit from minor performance improvements by decomposing some loops. For example in `solve_tril`:
```chapel 
for i in first..last {
    const sol = if unit_diag then y(i) else y(i) / L(i,i);
    y(i) = sol;

    if (i < last) {
      forall j in (i+1)..last with (ref y) {
        y(j) -= L(j,i) * sol;
      }
    }
  }
```
The `if`-statement is repeated each iteration, but is satisfied for only once (for `i = last`). This could cause performance degradation when the array size is large. We could rewrite this as:
```chapel 
for i in first..(last-1) {
    const sol = if unit_diag then y(i) else y(i) / L(i,i);
    y(i) = sol;

    forall j in (i+1)..last with (ref y) {
      y(j) -= L(j,i) * sol;
    }
  }
const sol = if unit_diag then y(last) else y(last) / L(last,last);
y(last) = sol;
```

**Review by jeremiah-corrado — CHANGES_REQUESTED** [review]
Hi @Guillaume-Helbecque, these changes look good!

Just one note about the `permute` procedure:

File: modules/packages/LinearAlgebra.chpl [hunk 1]
```diff
@@ -1645,9 +1649,8 @@ proc lu(A: [?Adom] ?eltType) {
 /* Return a new array as the permuted form of `A` according to
     permutation array `ipiv`.*/
 private proc permute(ipiv: [] int, A: [?Adom] ?eltType, transpose=false) {
-  const n = Adom.shape(0);
-  const dim = 0..<n;
-  var B: [Adom] eltType;
+  const dim = Adom;
+  var B: [dim] eltType;
```
[the remark this answers is not in the document]
Comment (jeremiah-corrado) [reply modules/packages/LinearAlgebra.chpl:1653]: Hmm, I didn't realize it was private and only used in the 1D case. I'd be inclined to remove the 2D branch of that procedure then (where we could always add it back later on if a use case arose).

**Review by jeremiah-corrado — APPROVED** [review]
Hi @Guillaume-Helbecque, looks good!

Let me run our test suite on your branch one more time and then I'll merge if there aren't any issues.

**Comment by Guillaume-Helbecque** [comment]
Hi @jeremiah-corrado, thanks for merging. Could you please edit the initial comment in #17624 in order to reflect this PR?

**Comment by jeremiah-corrado** [comment]
Sure, it's been updated. Thanks again for contributing!

**Pull request merged**